modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/report/BaseDDMFormFieldTypeReportProcessor.java

```diff
@@ -0,0 +1,181 @@
+
+		setIconAndLabelProperties(fieldJSONObject, ddmFormFieldTypeProperties);
+
+		fieldJSONObject.put(
+			"chartComponentName", getChartComponentName()
+		).put(
+			"chartComponentProps",
+			getChartComponentPropsJSONObject(fieldJSONObject, ddmFormFieldValue)
+		);
+
+		return fieldJSONObject;
+	}
+
+	protected abstract JSONObject doProcess(
+			DDMFormFieldValue ddmFormFieldValue, JSONObject fieldJSONObject,
+			long formInstanceRecordId, String ddmFormInstanceReportEvent)
+		throws Exception;
+
+	protected abstract String getChartComponentName();
+
+	protected abstract JSONObject getChartComponentPropsJSONObject(
+		JSONObject fieldJSONObject, DDMFormFieldValue ddmFormFieldValue);
+
+	protected JSONObject getDDMFormFieldOptionLabelsJSONObject(
+		DDMFormFieldOptions ddmFormFieldOptions) {
+
+		JSONObject jsonObject = JSONFactoryUtil.createJSONObject();
+
+		int index = 0;
+
+		for (String optionValue : ddmFormFieldOptions.getOptionsValues()) {
+			jsonObject.put(
+				optionValue,
+				JSONUtil.put(
+					"index", index++
+				).put(
+					"value",
+					getValue(ddmFormFieldOptions.getOptionLabels(optionValue))
+				));
+		}
+
+		return jsonObject;
+	}
+
+	protected String getValue(Value value) {
+		return value.getString(value.getDefaultLocale());
+	}
+
+	protected JSONArray mapToValueProperty(JSONArray jsonArray) {
+		JSONArray valuesJSONArray = JSONFactoryUtil.createJSONArray();
+
+		for (Object object : jsonArray) {
+			JSONObject valueJSONObject = (JSONObject)object;
+
+			valuesJSONArray.put(valueJSONObject.get("value"));
+		}
+
+		return valuesJSONArray;
+	}
+
+	protected void setIconAndLabelProperties(
+		JSONObject fieldJSONObject,
+		Map<String, Object> ddmFormFieldTypeProperties) {
+
+		fieldJSONObject.put(
+			"icon", ddmFormFieldTypeProperties.get("ddm.form.field.type.icon")
+		).put(
+			"title", ddmFormFieldTypeProperties.get("ddm.form.field.type.label")
+		);
+	}
+
+	protected int sumTotalValues(JSONObject valuesJSONObject) {
+		Stream<String> stream = valuesJSONObject.keySet(
+		).stream();
+
+		return stream.map(
+			key -> valuesJSONObject.getInt(key)
+		).reduce(
+			Integer::sum
+		).orElse(
+			0
+		);
+	}
+
+	protected JSONArray toDataArray(
+		DDMFormFieldOptions ddmFormFieldOptions, JSONObject valuesJSONObject) {
+
+		JSONArray dataJSONArray = JSONFactoryUtil.createJSONArray();
+
+		if (ddmFormFieldOptions == null) {
+			return dataJSONArray;
+		}
+
+		JSONObject optionsJSONObject = getDDMFormFieldOptionLabelsJSONObject(
+			ddmFormFieldOptions);
+
+		Stream<String> stream = valuesJSONObject.keySet(
+		).stream();
+
+		stream.map(
+			name -> {
+				int count = valuesJSONObject.getInt(name);
+
+				String label = name;
+
+				JSONObject optionDataJSONObject =
+					optionsJSONObject.getJSONObject(name);
+
+				if (optionDataJSONObject != null) {
+					label = optionDataJSONObject.getString("value");
+				}
+
+				return JSONUtil.put(
+					"count", count
+				).put(
+					"label", label
+				);
+			}
+		).sorted(
```

Comment:
Hey, @carolmariaabb . It will call `getDDMFormFieldOptionLablesJSONObject` multiple times with the same argument, so I'd declare and initialize `optionsJSONObject` it outside the `map` lambda. Also, I was trying to make the JSON object the most compact (to make it as close as possible to the JS notation), but it's preferable to use the Supplier, just let me know.

Reply:
Hi @cleydyr, you can do as I suggested, but declaring and initializing the `optionsJSONObject` outside the map lambda. But, If you prefer to keep it your way, please do `"count", valuesJSONObject.getInt(name)` instead of creating the `int count` variable. Thanks!

Reply:
Hi @cleydyr, you can do as I suggested, but declaring and initializing the optionsJSONObject outside the map lambda. But, If you prefer to keep it your way, please do "count", valuesJSONObject.getInt(name) instead of creating the int count variable. Thanks!